docs/configuration.md
# ⚙️ Configuring Sweep

Sweep can be configured through a `sweep.toml` file, which is read when running the binary, or through environment variables prefixed with `SWEEP__` and using `__` for splitting keys from field names (`llm.api_key` would become `SWEEP__LLM__API_KEY`).

> [!IMPORTANT]
> Fields marked with `[sensitive]` contain sensitive data. These should be set via environment variables. The environment variable name is shown next to the normal field name for such fields.

> [!NOTE]
> **Fields with no default value are required to be set by the user**.

Every heading in this guide gives an overview of a top-level key. Example: A heading has the top-level key `[approval]` and documents a field named `timeout` of type `unsigned 64-bit integer`. To configure that field, you'd write this into your `sweep.toml`:

```toml
# ONLY AN EXAMPLE!

[approval]
timeout = 45 # Now the timeout is 45 seconds!
```

## 🛡️ Approval Configuration `[approval]`

| Field             | Type                                          | Default Value | Description                                                                               |
| ----------------- | --------------------------------------------- | ------------- | ----------------------------------------------------------------------------------------- |
| `timeout`         | `unsigned 64-bit integer`                     | `30`          | The period of time (in seconds) in which an approval has to be answered before timing out |
| `skip_completion` | list of `"approved" \| "denied" \| "timeout"` | `[]`          | A list of approval outcomes that shouldn't trigger the LLM endpoint                       |

## 🤖 Bot Configuration `[bot]`

| Field                    | Type                      | Default Value | Description                                                                                                                              |
| ------------------------ | ------------------------- | ------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| `debounce_ms`            | `unsigned 64-bit integer` | `1000`        | How long the bot collects events to batch them together                                                                                  |
| `max_turns`              | `unsigned size`           | `10`          | The number of consecutive tool call turns a model can take before being interrupted. Setting this to 0 will disable tools **everywhere** |
| `wake_on_mention`        | `bool`                    | `true`        | Whether Sweep should only respond when explicitly mentioned                                                                              |
| `wake_on_mention_notify` | `bool`                    | `true`        | Whether to send a notification if the `wake_on_mention` mode is enabled                                                                  |

## 💬 Channel Configuration `[channel]`

> [!NOTE]
> Setting both `channel.blacklist` and `channel.whitelist` will produce a warning. The blacklist would be ignored in such case.

| Field       | Type                              | Default Value | Description                                                                   |
| ----------- | --------------------------------- | ------------- | ----------------------------------------------------------------------------- |
| `blacklist` | list of `unsigned 64-bit integer` | `[]`          | A list of channel IDs that Sweep **will ignore**                              |
| `whitelist` | list of `unsigned 64-bit integer` | `[]`          | A list of channel IDs to restrict Sweep's usage to **only those channel IDs** |

### 📝 Channel-specific Configuration `[[channel.override]]`

A channel override can be used for configuring Sweep's behavior in special channels. Overrides also take priority over a normal whitelist/blacklist.

| Field               | Type                         | Default Value | Description                                                                                                  |
| ------------------- | ---------------------------- | ------------- | ------------------------------------------------------------------------------------------------------------ |
| `id`                | `unsigned 64-bit integer`    | **None**      | The ID of the channel                                                                                        |
| `enable`            | `bool`                       | **None**      | Whether Sweep can be used in this channel                                                                    |
| `disable_all_tools` | `bool`                       | `false`       | Whether all tools should be disabled in this channel. Takes priority over `enable_tools` and `disable_tools` |
| `enable_tools`      | list of `internal tool name` | `[]`          | If set, all tools will be disabled, except these                                                             |
| `disable_tools`     | list of `internal tool name` | `[]`          | If set, all tools will be enabled, except these                                                              |

#### Example

```toml
[[channel.override]]
id = 1234
enable = true
enable_tools = ["create_category"]

[[channel.override]]
id = 5678
enable = true
disable_all_tools = true

[[channel.override]]
id = 9101
enable = true
disable_tools = ["get_local_time"]
```

#### Using Environment Variables

You can also use environment variables for setting channel-specific configurations:

```
SWEEP__CHANNEL__OVERRIDE='[{id=1234,enable=false},{id=5678,enable=true,disable_all_tools=true}]'
```

## 🔷 Discord Configuration `[discord]`

| Field                                             | Type     | Default Value | Description           |
| ------------------------------------------------- | -------- | ------------- | --------------------- |
| **[sensitive]** `token` (`SWEEP__DISCORD__TOKEN`) | `string` | **None**      | The Discord bot token |

## 🌐 LLM Endpoint Configuration `[llm]`

| Field                                             | Type     | Default Value | Description                                                                                   |
| ------------------------------------------------- | -------- | ------------- | --------------------------------------------------------------------------------------------- |
| `endpoint`                                        | `string` | **None**      | The base URL of your OpenAI-compatible endpoint. **Make sure it doesn't have a trailing `/`** |
| `model`                                           | `string` | **None**      | The model you are using. You may use `""` for local backends that only have one loaded model  |
| **[sensitive]** `api_key` (`SWEEP__LLM__API_KEY`) | `string` | `""`          | Your API-key for your provider                                                                |
| `project_id`                                      | `string` | `""`          | The project ID of your application                                                            |
| `org_id`                                          | `string` | `""`          | The organization ID of your application                                                       |

## 🛠️ Tool Configuration `[tools]`

| Field     | Type                         | Default Value | Description                       |
| --------- | ---------------------------- | ------------- | --------------------------------- |
| `disable` | list of `internal tool name` | `[]`          | A list of globally disabled tools |

## 👥 User Configuration `[users]`

> [!NOTE]
> Setting both `users.blacklist` and `users.whitelist` will produce a warning. The blacklist would be ignored in such case.

| Field       | Type                              | Default Value | Description                                               |
| ----------- | --------------------------------- | ------------- | --------------------------------------------------------- |
| `blacklist` | list of `unsigned 64-bit integer` | `[]`          | A list of user IDs that Sweep **will ignore**             |
| `whitelist` | list of `unsigned 64-bit integer` | `[]`          | A list of user IDs of people who are allowed to use Sweep |


src/config.rs
use std::fmt::Display;

use figment::{
    Figment,
    providers::{Env, Format, Toml},
};
use serde::Deserialize;

use crate::core::access_filter::AccessFilter;

#[derive(Deserialize)]
pub struct Configuration {
    #[serde(default)]
    pub approval: ApprovalConfig,
    #[serde(default)]
    pub bot: BotConfig,
    #[serde(default)]
    pub channel: ChannelConfig,

    pub discord: DiscordConfig,
    pub llm: LlmConfig,

    #[serde(default)]
    pub tools: ToolsConfig,
    #[serde(default)]
    pub users: UsersConfig,
}

#[derive(Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SkippedCompletionEvent {
    Approved,
    Denied,
    Timeout,
}

impl Display for SkippedCompletionEvent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let name = match self {
            Self::Approved => "approved",
            Self::Denied => "denied",
            Self::Timeout => "timeout",
        };

        write!(f, "{}", name)
    }
}

#[derive(Deserialize)]
#[serde(default)]
pub struct ApprovalConfig {
    pub timeout: u64,
    pub skip_completion: Vec<SkippedCompletionEvent>,
}

impl Default for ApprovalConfig {
    fn default() -> Self {
        Self {
            timeout: 30,
            skip_completion: Vec::new(),
        }
    }
}

#[derive(Deserialize)]
#[serde(default)]
pub struct BotConfig {
    pub debounce_ms: u64,
    pub max_turns: usize,
    pub wake_on_mention: bool,
    pub wake_on_mention_notify: bool,
}

impl Default for BotConfig {
    fn default() -> Self {
        Self {
            debounce_ms: 1000,
            max_turns: 10,
            wake_on_mention: true,
            wake_on_mention_notify: true,
        }
    }
}

#[derive(Deserialize, Default)]
#[serde(default)]
pub struct ChannelConfig {
    #[serde(flatten)]
    pub access_filter: AccessFilter<u64>,
    #[serde(rename = "override")]
    pub overrides: Vec<ChannelOverride>,
}

#[derive(Deserialize)]
pub struct ChannelOverride {
    pub id: u64,
    pub enable: bool,
    #[serde(default)]
    pub disable_all_tools: bool,
    #[serde(default)]
    pub disable_tools: Vec<String>,
    #[serde(default)]
    pub enable_tools: Vec<String>,
}

#[derive(Deserialize)]
pub struct DiscordConfig {
    pub token: String,
}

#[derive(Deserialize)]
pub struct LlmConfig {
    pub model: String,
    pub endpoint: String,

    #[serde(default)]
    pub api_key: Option<String>,
    #[serde(default)]
    pub project_id: Option<String>,
    #[serde(default)]
    pub org_id: Option<String>,
}

#[derive(Deserialize, Default)]
pub struct ToolsConfig {
    pub disable: Vec<String>,
}

#[derive(Deserialize, Default)]
pub struct UsersConfig {
    #[serde(flatten)]
    pub access_filter: AccessFilter<u64>,
}

// We ignore the warning, because we need the error variant
#[allow(clippy::result_large_err)]
pub fn load() -> figment::Result<Configuration> {
    Figment::new()
        .merge(Toml::file("sweep.toml"))
        .join(Env::prefixed("SWEEP__").split("__"))
        .extract()
}
